### DISCUSSION TITLE
How to prevent randomness when converting images?

### DISCUSSION THREAD

**Discussion by peterdell** [ImageMagick/ImageMagick#7617] [Help] [answered]
### ImageMagick version

ImageMagick-7.1.0-Q16-HDRI

### Operating system, version and so on

Windows 10

### Description

Hello,

I use magick.exe in a batch script to crop/resize/combine nine single pictures into one in 3 resolutions/formats. Technically, it all works great. Here is the full batch script [Make-Images.bat](https://github.com/wudsn-atari800-demos/SillyPack2k22SE/blob/main/Make-Images.bat) for your reference.

But now that I've switched to git for versioning, I see that the apparently same result is different on the binary level with every run. This creates issues with git. It looks like ImageMagick is applying some randomness during the conversion.

Is there an option to suppress this or explicitly specify the seed for the randomness to achieve reproducible results?

Kind regards, Peter.

**Comment by snibgo** [comment]
I can't see any randomness in your script. If you run it twice with the same inputs, you should get the same output images.

[On setting the seed for pseudo-random numbers, see https://imagemagick.org/script/command-line-options.php#seed .]

Some image file formats contain date/time stamps. This is true of PNG files. You can exclude the date/time stamps when writing PNG files like this:
```
magick toes.png -define png:exclude-chunk=date x0.png
```
See https://imagemagick.org/script/defines.php

Incidentally, many of your multiple `magick` commands could be combined, which would also reduce the need to write and read intermediate results.

**Comment by peterdell** [comment, marked as answer]
Thanks a lot for the quick reply. As it turns out, there really is no randomness (this was only displayed as pixel difference in GitHub Desktop, which is instead an issue in that tool). In addition to "date" also the ExIf chunk was an issue. 

Answer: Using **"-strip"** as the last option before the .png output file solves my issue.

I would appreciate your suggestions on how to combine the commands. I assume you refer to the "combine" lines, but I can't find an example.

**Comment by snibgo** [comment]
First, I suggest changing each "`magick composite`" to just "`magick`" with "`-composite`". You need to reverse the order of the input files, and put "`-geometry`" after the inputs. So:
```
%MAGICK% images\Template.png %TEMP%\1.png -geometry +0+0 -composite %TEMP_PNG%
%MAGICK% %TEMP_PNG% %TEMP%\2.png -geometry +337+0 -composite %TEMP_PNG%
%MAGICK% %TEMP_PNG% %TEMP%\3.png -geometry +674+0 -composite %TEMP_PNG%
:
%MAGICK% %TEMP_PNG% %TEMP%\9.png -geometry +674+443 -composite %TEMP_PNG%
```
... and so on. Then we can combine those commands. This is rather like adding two numbers to make a result, then adding the result to another number, and so on.
```
%MAGICK% images\Template.png ^
  %TEMP%\1.png -geometry +0+0 -composite ^
  %TEMP%\2.png -geometry +337+0 -composite ^
  %TEMP%\3.png -geometry +674+0 -composite ^
  :
  %TEMP%\9.png -geometry +674+443 -composite %TEMP_PNG%
```
Check that these changes work. Then, you can substitute each "`%TEMP%\1.png`" with the image expression that created it. Put that expression in parentheses "()". I suggest inserting "`+repage`" after each "`-crop`".
```
%MAGICK% images\Template.png ^
  ( images\1.png -scale 336x240 -crop 336x200+0+17 +repage ) -geometry +0+0 -composite ^
  ( images\2.png -scale 336x240 -crop 336x200+0+20 +repage ) -geometry +337+0 -composite ^
  ( images\3.png -scale 336x240 -crop 336x200+0+20 +repage ) -geometry +674+0 -composite ^
  :
  ( images\9.png -scale 336x240 ) -geometry +674+443 -composite %TEMP_PNG%
```
Again, check that these changes work.

In this way, your script can become a single `magick` command, with no writes or reads of temporary files.

**Comment by peterdell** [comment]
Excellent, it works as intended. Thank you a lot for these quick and to-the-point answers!

https://github.com/wudsn-atari800-demos/SillyPack2k22SE/commit/93342720a37fdcf6895fd43830692d3e0cb63fef
